File: backend/documents-service/src/main/java/com/quisin/document/controller/DocumentController.java
package com.quisin.document.controller;

import com.quisin.document.dto.DocumentRequest;
import com.quisin.document.dto.DocumentResponse;
import com.quisin.document.model.DocumentType;
import com.quisin.document.service.DocumentService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/v1/documents")
public class DocumentController {

    private final DocumentService documentService;

    @PostMapping("/generate")
    public ResponseEntity<byte[]> generateDocument(@Valid @RequestBody DocumentRequest request) {
        DocumentResponse response = documentService.generateDocument(request);

        return ResponseEntity.ok()
                .contentType(MediaType.parseMediaType(response.getContentType()))
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + response.getFileName() + "\"")
                .body(response.getContent());
    }

    @GetMapping("/qrcode/{restaurantId}/{tableNumber}")
    public ResponseEntity<byte[]> generateQRCode(
            @PathVariable String restaurantId,
            @PathVariable String tableNumber,
            @RequestParam(defaultValue = "PNG") String format) {

        DocumentRequest request = DocumentRequest.builder()
                .restaurantId(restaurantId)
                .tableNumber(tableNumber)
                .documentType(DocumentType.QR_CODE)
                .format(format)
                .build();

        DocumentResponse response = documentService.generateDocument(request);

        return ResponseEntity.ok()
                .contentType(MediaType.parseMediaType(response.getContentType()))
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + response.getFileName() + "\"")
                .body(response.getContent());
    }

    @GetMapping("/report/{restaurantId}/{type}")
    public ResponseEntity<byte[]> generateReport(
            @PathVariable String restaurantId,
            @PathVariable DocumentType type,
            @RequestParam(required = false) String startDate,
            @RequestParam(required = false) String endDate,
            @RequestParam(defaultValue = "PDF") String format) {

        DocumentRequest request = DocumentRequest.builder()
                .restaurantId(restaurantId)
                .documentType(type)
                .startDate(startDate)
                .endDate(endDate)
                .format(format)
                .build();

        DocumentResponse response = documentService.generateDocument(request);

        return ResponseEntity.ok()
                .contentType(MediaType.parseMediaType(response.getContentType()))
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + response.getFileName() + "\"")
                .body(response.getContent());
    }

    @GetMapping("/credentials/{restaurantId}")
    public ResponseEntity<byte[]> generateRestaurantCredentials(@PathVariable String restaurantId) {
        byte[] content = documentService.generateRestaurantCredentials(restaurantId);

        return ResponseEntity.ok()
                .contentType(MediaType.APPLICATION_PDF)
                .header(HttpHeaders.CONTENT_DISPOSITION, 
                        "attachment; filename=\"restaurant_credentials_" + restaurantId + ".pdf\"")
                .body(content);
    }

    @GetMapping("/audit/{restaurantId}")
    public ResponseEntity<byte[]> generateAuditReport(
            @PathVariable String restaurantId,
            @RequestParam String startDate,
            @RequestParam String endDate) {
        
        byte[] content = documentService.generateAuditReport(restaurantId, startDate, endDate);

        return ResponseEntity.ok()
                .contentType(MediaType.APPLICATION_PDF)
                .header(HttpHeaders.CONTENT_DISPOSITION, 
                        "attachment; filename=\"audit_report_" + restaurantId + ".pdf\"")
                .body(content);
    }

    @GetMapping("/analytics/{restaurantId}")
    public ResponseEntity<byte[]> generateAnalyticsReport(
            @PathVariable String restaurantId,
            @RequestParam String startDate,
            @RequestParam String endDate,
            @RequestParam(defaultValue = "PDF") String format) {
        
        byte[] content = documentService.generateAnalyticsReport(restaurantId, startDate, endDate, format);
        String extension = format.toLowerCase();
        String contentType = switch (format.toUpperCase()) {
            case "PDF" -> MediaType.APPLICATION_PDF_VALUE;
            case "EXCEL" -> "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet";
            case "CSV" -> "text/csv";
            default -> throw new IllegalArgumentException("Unsupported format: " + format);
        };

        return ResponseEntity.ok()
                .contentType(MediaType.parseMediaType(contentType))
                .header(HttpHeaders.CONTENT_DISPOSITION, 
                        "attachment; filename=\"analytics_report_" + restaurantId + "." + extension + "\"")
                .body(content);
    }

    @GetMapping("/performance/{restaurantId}")
    public ResponseEntity<byte[]> generatePerformanceReport(
            @PathVariable String restaurantId,
            @RequestParam(defaultValue = "DAILY") String period) {
        
        byte[] content = documentService.generatePerformanceReport(restaurantId, period);

        return ResponseEntity.ok()
                .contentType(MediaType.APPLICATION_PDF)
                .header(HttpHeaders.CONTENT_DISPOSITION, 
                        "attachment; filename=\"performance_report_" + restaurantId + ".pdf\"")
                .body(content);
    }

    @GetMapping("/inventory/{restaurantId}")
    public ResponseEntity<byte[]> generateInventoryReport(@PathVariable String restaurantId) {
        byte[] content = documentService.generateInventoryReport(restaurantId);

        return ResponseEntity.ok()
                .contentType(MediaType.APPLICATION_PDF)
                .header(HttpHeaders.CONTENT_DISPOSITION, 
                        "attachment; filename=\"inventory_report_" + restaurantId + ".pdf\"")
                .body(content);
    }

    @GetMapping("/tables/{restaurantId}")
    public ResponseEntity<byte[]> generateTableReport(@PathVariable String restaurantId) {
        byte[] content = documentService.generateTableReport(restaurantId);

        return ResponseEntity.ok()
                .contentType(MediaType.APPLICATION_PDF)
                .header(HttpHeaders.CONTENT_DISPOSITION, 
                        "attachment; filename=\"table_report_" + restaurantId + ".pdf\"")
                .body(content);
    }

    @GetMapping("/menu/{restaurantId}")
    public ResponseEntity<byte[]> generateMenuReport(@PathVariable String restaurantId) {
        byte[] content = documentService.generateMenuReport(restaurantId);

        return ResponseEntity.ok()
                .contentType(MediaType.APPLICATION_PDF)
                .header(HttpHeaders.CONTENT_DISPOSITION, 
                        "attachment; filename=\"menu_report_" + restaurantId + ".pdf\"")
                .body(content);
    }

    @GetMapping("/staff/credentials/{restaurantId}/{staffId}")
    public ResponseEntity<byte[]> generateStaffCredentials(
            @PathVariable String restaurantId,
            @PathVariable String staffId,
            @RequestParam String username,
            @RequestParam String password) {
        
        byte[] content = documentService.generateStaffCredentials(restaurantId, staffId, username, password);

        return ResponseEntity.ok()
                .contentType(MediaType.APPLICATION_PDF)
                .header(HttpHeaders.CONTENT_DISPOSITION, 
                        "attachment; filename=\"staff_credentials_" + staffId + ".pdf\"")
                .body(content);
    }
} 

File: backend/documents-service/src/main/kotlin/com/quisin/documents/controller/DocumentController.kt
package com.quisin.documents.controller

import com.quisin.documents.service.DocumentGenerationService
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.math.BigDecimal
import java.time.LocalDateTime

@RestController
@RequestMapping("/api/documents")
class DocumentController(
    private val documentGenerationService: DocumentGenerationService
) {
    @PostMapping("/invoice/pdf")
    fun generateInvoicePDF(
        @RequestBody invoiceRequest: InvoiceRequest
    ): ResponseEntity<ByteArray> {
        val invoiceData = DocumentGenerationService.InvoiceData(
            invoiceNumber = invoiceRequest.invoiceNumber,
            customerName = invoiceRequest.customerName,
            orderDate = invoiceRequest.orderDate,
            items = invoiceRequest.items.map { 
                DocumentGenerationService.InvoiceItem(
                    name = it.name,
                    quantity = it.quantity,
                    unitPrice = it.unitPrice
                )
            },
            totalAmount = invoiceRequest.items.sumOf { it.quantity * it.unitPrice }
        )

        val pdfBytes = documentGenerationService.generateInvoicePDF(invoiceData)

        val headers = HttpHeaders().apply {
            contentType = MediaType.APPLICATION_PDF
            contentDisposition = org.springframework.http.ContentDisposition.builder("attachment")
                .filename("invoice_${invoiceData.invoiceNumber}.pdf")
                .build()
        }

        return ResponseEntity(pdfBytes, headers, HttpStatus.OK)
    }

    @PostMapping("/report/excel")
    fun generateExcelReport(
        @RequestBody reportData: List<Map<String, Any>>
    ): ResponseEntity<ByteArray> {
        val excelBytes = documentGenerationService.generateExcelReport(reportData)

        val headers = HttpHeaders().apply {
            contentType = MediaType.parseMediaType("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet")
            contentDisposition = org.springframework.http.ContentDisposition.builder("attachment")
                .filename("quisin_report_${System.currentTimeMillis()}.xlsx")
                .build()
        }

        return ResponseEntity(excelBytes, headers, HttpStatus.OK)
    }

    @PostMapping("/report/csv")
    fun generateCSVReport(
        @RequestBody reportData: List<Map<String, Any>>
    ): ResponseEntity<ByteArray> {
        val csvBytes = documentGenerationService.generateCSVReport(reportData)

        val headers = HttpHeaders().apply {
            contentType = MediaType.parseMediaType("text/csv")
            contentDisposition = org.springframework.http.ContentDisposition.builder("attachment")
                .filename("quisin_report_${System.currentTimeMillis()}.csv")
                .build()
        }

        return ResponseEntity(csvBytes, headers, HttpStatus.OK)
    }

    // Data Transfer Objects
    data class InvoiceRequest(
        val invoiceNumber: String,
        val customerName: String,
        val orderDate: LocalDateTime,
        val items: List<InvoiceItemRequest>
    )

    data class InvoiceItemRequest(
        val name: String,
        val quantity: Int,
        val unitPrice: BigDecimal
    )
}
